File: 2021_summer_algorithm_study_intermediate/using_cpp/2150_kosaraju.cpp
/*
input :
7 9
1 4
4 5
5 1
1 6
6 7
2 7
7 3
3 7
7 2

output :
3
1 4 5 -1
2 3 7 -1
6 -1
*/

#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

// 전역 변수 선언
int V, E;
int a, b;
int SCC_NUM;
vector<int> graph[10005]; // 10005
vector<int> graph_inv[10005]; // 10005
bool visited[10005]; // 10005
int scc[10005]; // 10005
vector<int> st;
vector<int> ans[10005]; // 10005

bool comp(vector<int> a, vector<int> b){
    return a[0] < b[0];
}

void dfs1(int s){
    visited[s] = true;
    for (int nxt: graph[s]){
        if (visited[nxt]) continue;

        dfs1(nxt);
    }
    st.push_back(s);
}

void dfs2(int s){
    scc[s] = SCC_NUM;

    ans[SCC_NUM].push_back(s);
    for (int nxt: graph_inv[s]){
        if (scc[nxt]) continue;

        dfs2(nxt);
    }
}

int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> V >> E;
    for (int i = 0; i < E; i++){
        cin >> a >> b;
        graph[a].push_back(b);
        graph_inv[b].push_back(a);
    }

    for (int i = 1; i <= V; i++){
        if (visited[i]) continue;

        dfs1(i);
    }

    while(!st.empty()){
        int p = st.back(); st.pop_back();
        if (scc[p]) continue;

        SCC_NUM++;
        dfs2(p);
        sort(ans[SCC_NUM].begin(), ans[SCC_NUM].end());
    }
    sort(ans + 1, ans + SCC_NUM + 1, comp);

    cout << SCC_NUM << '\n';
    for (int i = 1; i <= SCC_NUM; i++){
        for (int j = 0; j < ans[i].size(); j++){
            cout << ans[i][j] << ' ';
        }
        cout << "-1\n";
    }

    return 0;
}

File: 2021_summer_algorithm_study_intermediate/using_python/2150_Kosaraju.py
"""
input :
7 9
1 4
4 5
5 1
1 6
6 7
2 7
7 3
3 7
7 2

output :
3
1 4 5 -1
2 3 7 -1
6 -1
"""
import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline
sys.setrecursionlimit(10 ** 6)


def dfs1(s):
    visited[s] = True
    for nxt in graph[s]:
        if visited[nxt]: continue

        dfs1(nxt)

    st.append(s)


def dfs2(s):
    scc[s] = SCC_NUM
    for nxt in graph_inv[s]:
        if scc[nxt] != 0: continue

        dfs2(nxt)


if __name__ == "__main__":
    V, E = map(int, input().rstrip().split())
    graph = [[] for _ in range(V + 1)]
    graph_inv = [[] for _ in range(V + 1)]
    for _ in range(E):
        a, b = map(int, input().rstrip().split())
        graph[a].append(b)
        graph_inv[b].append(a)

    visited = [False for _ in range(V + 1)]
    st = []
    for i in range(1, V + 1):
        if visited[i]: continue

        dfs1(i)

    scc = [0 for _ in range(V + 1)]
    SCC_NUM = 0
    while len(st) != 0:
        p = st.pop()
        if scc[p] != 0: continue

        SCC_NUM += 1
        dfs2(p)

    res = [[] for _ in range(V + 1)]
    for i in range(1, V + 1):
        res[scc[i]].append(i)
    res.sort()

    print(SCC_NUM)
    for group in res:
        if group == []: continue
        for node in group:
            print(node, end=' ')
        print(-1)
